File: app/src/main/java/com/arejas/dashboardofthings/domain/entities/result/LiveDataResource.java
package com.arejas.dashboardofthings.domain.entities.result;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import com.arejas.dashboardofthings.utils.functional.Consumer;
import com.arejas.dashboardofthings.utils.functional.Supplier;

public class LiveDataResource<T> extends MediatorLiveData<Resource<T>> {

    public LiveDataResource(Supplier<LiveData<T>> dataSupplier) {
        try {
            LiveData<T> data = dataSupplier.get();
            addSource(data, received -> {
                if (received == null) {
                    postValue(Resource.loading(null));
                } else {
                    postValue(Resource.success(received));
                }
            });
        } catch (Exception e) {
            postValue(Resource.error(e, null));
        }
    }

    public LiveDataResource(Supplier<LiveData<T>> dataSupplier, Consumer<T> dataModifier) {
        try {
            LiveData<T> data = dataSupplier.get();
            addSource(data, received -> {
                try {
                    if (received == null) {
                        postValue(Resource.loading(null));
                    } else {
                        dataModifier.accept(received);
                        postValue(Resource.success(received));
                    }
                } catch (Exception e) {
                    postValue(Resource.error(e, null));
                }
            });
        } catch (Exception e) {
            postValue(Resource.error(e, null));
        }
    }
}

File: app/src/main/java/com/arejas/dashboardofthings/domain/entities/result/LiveDataResource.kt
package com.arejas.dashboardofthings.domain.entities.result

import androidx.core.util.Consumer
import androidx.core.util.Supplier
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData

class LiveDataResource<T> : MediatorLiveData<Resource<T>> {

    constructor(dataSupplier: () -> LiveData<T> ) {
        try {
            val data = dataSupplier()
            addSource(data) { received ->
                if (received == null) {
                    postValue(Resource.loading(null))
                } else {
                    postValue(Resource.success(received))
                }
            }
        } catch (e: Exception) {
            postValue(Resource.error(e, null))
        }

    }

    constructor(dataSupplier: Supplier<LiveData<T>>, dataModifier: Consumer<T>) {
        try {
            val data = dataSupplier.get()
            addSource(data) { received ->
                try {
                    if (received == null) {
                        postValue(Resource.loading(null))
                    } else {
                        dataModifier.accept(received)
                        postValue(Resource.success(received))
                    }
                } catch (e: Exception) {
                    postValue(Resource.error(e, null))
                }
            }
        } catch (e: Exception) {
            postValue(Resource.error(e, null))
        }

    }
}
